cpp/src/arrow/testing/random.cc

```diff
@@ -608,6 +609,218 @@ std::shared_ptr<Array> OffsetsFromLengthsArray(OffsetArrayType* lengths,
+///
+/// When sparsity is 0.0, the list-view spans are perfectly packed one after the
+/// other. If sparsity is greater than 0.0, the list-view spans are set apart
+/// from each other in proportion to the sparsity value and size of each
+/// list-view. A negative sparsity means each list-view shares a fraction of the
+/// values used by the previous list-view.
+///
+/// For instance, a sparsity of -1.0 means the values array will only need enough values
+/// for the largest list-view with all the other list-views spanning some of these same
+/// values.
+///
+/// \param[in] seed The seed for the random number generator
+/// \param[in,out] mutable_sizes_array The array of sizes to use
+/// \param[in] force_empty_nulls Whether to force null list-view sizes to be 0
+/// \param[in] zero_undefined_offsets Whether to zero the offsets of list-views that have
+/// 0 set as the size
+/// \param[in] sparsity The sparsity of the generated list-view offsets
+/// \param[out] out_max_view_end The maximum value of the end of a list-view
+template <typename OffsetArrayType, typename offset_type>
+std::shared_ptr<Array> ViewOffsetsFromLengthsArray(
+    SeedType seed, OffsetArrayType& mutable_sizes_array, bool force_empty_nulls,
+    bool zero_undefined_offsets, double sparsity, int64_t* out_max_view_end,
+    int64_t alignment, MemoryPool* memory_pool) {
+  using TypeClass = typename OffsetArrayType::TypeClass;
+
+  auto* sizes = mutable_sizes_array.data()->template GetMutableValues<offset_type>(1);
+
+  BufferVector buffers{2};
+  buffers[0] = NULLPTR;  // sizes can have nulls, offsets don't have to
+  buffers[1] = *AllocateBuffer(sizeof(offset_type) * mutable_sizes_array.length(),
+                               alignment, memory_pool);
+  auto offsets = buffers[1]->mutable_data_as<offset_type>();
+
+  double offset_base = 0.0;
+  offset_type max_view_end = 0;
+  for (int64_t i = 0; i < mutable_sizes_array.length(); ++i) {
+    const auto offset = static_cast<offset_type>(std::llround(offset_base));
+    if (mutable_sizes_array.IsNull(i)) {
+      if (force_empty_nulls) {
+        sizes[i] = 0;
+      }
+      offsets[i] = zero_undefined_offsets ? 0 : offset;
+    } else {
+      if (sizes[i] == 0) {
+        offsets[i] = zero_undefined_offsets ? 0 : offset;
+      } else {
+        offsets[i] = offset;
+        DCHECK_LT(offset, std::numeric_limits<offset_type>::max() - sizes[i]);
+        offset_base = std::max(0.0, offset_base + (sparsity * sizes[i]));
+      }
+    }
+    max_view_end = std::max(max_view_end, offsets[i] + sizes[i]);
+  }
+  *out_max_view_end = max_view_end;
+
+  auto array_data =
+      ArrayData::Make(TypeTraits<TypeClass>::type_singleton(),
+                      mutable_sizes_array.length(), std::move(buffers), /*null_count=*/0);
+  return std::make_shared<OffsetArrayType>(std::move(array_data));
+}
+
+template <typename ArrayType, typename RAG>
+Result<std::shared_ptr<Array>> ArrayOfListView(RAG& self, const Field& field,
```

Comment:
Hmm... why do `RandomListView` and `ArrayOfListView` use two different approaches? This is more code to maintain and understand.

(I'm not sure which approach is better, but ideally we would decide on a common one)

Reply:
It is indeed a problem, but note that regular list generators are also split in 2 versions. I started this PR with an unified version, but the code was too complicated (the parameterization is different for some reason) and after your first review I added the `random.cc: Simplify and split the random generator into two algorithms` commit.

I mentioned this duplication in the issues about `concatenate.cc` also having yet another random generator: https://github.com/apache/arrow/issues/38656#issuecomment-1806507591

Reply:
Could these at least use a similar `coverage` parameter? I don't think `sparsity` brings much to the table, and the fact that its description needs 8 lines of text probably means it won't get much use in practice.

Reply:
I was planning to use the sparsity concept in benchmarks as the tighter packing of the data could affect performance, but I will remove it for now and stash the code somewhere.